### PR TITLE
Objective outputs should rescale based on sample_weights

### PR DESCRIPTION
**Pull request by wxs**
If `sample_weight` is to be used as a mask as well as for re-weighting then it's important that, at least when used as a mask, the output be rescaled. Otherwise the order of magnitude of your objective changes purely based on the number of masked entries in your training data.

The exact implementation is perhaps confusing, and is due to overloading of `sample_weight` to be both a weight and a mask. I personally only use it as a mask, so perhaps anyone using it for weighting should make sure that this is appropriate for their use as well. It is because of the weighting use-case that, rather than simply divide by the mean, I first convert it to a binary mask and then divide by the mean.

@EderSantana you mentioned in #1520 that, with sample_weights, on new Keras you were getting smaller scores than you did before the backend change. This could be the source of that issue, as it would cause your "correct" loss to be multiplied by the ratio of masked to unmasked entries.

### CODE REVIEW

**Comment by fchollet** [comment]
Tests are failing, but it's not clear why. Let's restart.
